src/analysis/special_functions/exp_log.lean

```diff
@@ -449,4 +450,85 @@ lemma tendsto_pow_mul_exp_neg_at_top_nhds_0 (n : ℕ) : tendsto (λx, x^n * exp
 (tendsto_inv_at_top_zero.comp (tendsto_exp_div_pow_at_top n)).congr $ λx,
   by rw [function.comp_app, inv_eq_one_div, div_div_eq_mul_div, one_mul, div_eq_mul_inv, exp_neg]
 
+open_locale big_operators
+
+/-- A crude lemma estimating the difference between `log (1-x)` and its Taylor series at `0`,
+where the main point of the bound is that it tends to `0`. The goal is to deduce the series
+expansion of the logarithm, in `has_sum_pow_div_log_of_abs_lt_1`.
+-/
+lemma abs_log_sub_add_sum_range_le {x : ℝ} (h : abs x < 1) (n : ℕ) :
+  abs ((∑ i in range n, x^(i+1)/(i+1)) + log (1-x)) ≤ (abs x)^(n+1) / (1 - abs x) :=
+begin
+  /- For the proof, we show that the derivative of the function to be estimated is small,
+  and then apply the mean value inequality. -/
+  let F : ℝ → ℝ := λ x, ∑ i in range n, x^(i+1)/(i+1) + log (1-x),
+  -- First step: compute the derivative of `F`
+  have A : ∀ y ∈ set.Ioo (-1 : ℝ) 1, deriv F y = - (y^n) / (1 - y),
+  { assume y hy,
+    have : (∑ i in range n, (↑i + 1) * y ^ i / (↑i + 1)) = (∑ i in range n, y ^ i),
+    { congr,
+      ext i,
+      have : (i : ℝ) + 1 ≠ 0 := ne_of_gt (nat.cast_add_one_pos i),
+      field_simp [this, mul_comm] },
+    field_simp [F, this, ← geom_series_def, geom_sum (ne_of_lt hy.2),
+                sub_ne_zero_of_ne (ne_of_gt hy.2), sub_ne_zero_of_ne (ne_of_lt hy.2)],
+    ring },
+  -- second step: show that the derivative of `F` is small
+  have B : ∀ y ∈ set.Icc (-abs x) (abs x), abs (deriv F y) ≤ (abs x)^n / (1 - abs x),
+  { assume y hy,
+    have : y ∈ set.Ioo (-(1 : ℝ)) 1 := ⟨lt_of_lt_of_le (neg_lt_neg h) hy.1, lt_of_le_of_lt hy.2 h⟩,
+    calc abs (deriv F y) = abs (-(y^n) / (1 - y)) : by rw [A y this]
+    ... ≤ (abs x)^n / (1 - abs x) :
+      begin
+        simp only [← pow_abs, abs_div, abs_neg],
+        apply_rules [div_le_div, pow_nonneg, abs_nonneg, pow_le_pow_of_le_left],
+        show abs y ≤ abs x, from abs_le_of_le_of_neg_le hy.2 (by linarith [hy.1]),
+        show 0 < 1 - abs x, by linarith,
+        show 1 - abs x ≤ abs (1 - y), from le_trans (by linarith [hy.2]) (le_abs_self _)
```

Comment:
Are these three lines handling side goals from `apply_rules`? I'd usually wrap each in `{ }` to convey that (still on a single line each).

Reply:
Yes, they are leftover goals from `apply_rules`. The order of the goals is not obvious (it depends on the order in which `apply_rules` will apply the lemmas), so I'd rather not use `{ ... }` as this requires a specific order of the goals, contrary to the `show` idiom, which shows that you are closing a remaining goal but does not depend on the order of the goals. Another option is to replace the `show ...` with `have : ...` before the `apply_rules` (as `apply_rules` tries to close the goals with `assumption`), but it means that if `apply_rules` becomes more powerful the useless lines will remain. What do you think?

Reply:
I suggest
```
show foo, {exact term}
show bar, {linarith}
show baz, {exact term}
```
I usually use this style when the initial tactic is doing goal selection or manipulation and the rest acts as normal (for example `swap n` might also be in that position).

Reply:
But given that you are stating the types of these goals, it may make more sense to just use `have` as you propose. I'm not sure "if `apply_rules` becomes more powerful the useless lines will remain" is necessarily a bad thing; turning it around you are saying that this is a brittle proof that will break if a tactic that is known to get stronger on its own does so. I would rather discover this sort of thing with a lint while having the proofs still work.

Reply:
OK, changed to `have : ...` before `apply_rules`. This should be robust.

Reply:
I see, I wasn't familiar with using `show` for goal selection (although I'm glad to learn this idiom!) so I think I would have been puzzled by the old proof if it broke and I had to fix it.

Reply:
That's a very Isabelle-like way of presenting things (more readable than the usual Lean way, but more verbose, and possibly more robust).